File: packages/guards/src/language.ts
import type {
  Guard,
  GuardResult,
  GuardContext,
} from '@open-guardrail/core';

interface LanguageOptions {
  allowed: string[];
  action: 'block' | 'warn';
}

const COMMON_WORDS: Record<string, string[]> = {
  en: [
    'the', 'is', 'are', 'was', 'were',
    'have', 'has', 'this', 'that', 'with',
    'and', 'for', 'not', 'you', 'but',
  ],
  es: [
    'el', 'la', 'los', 'las', 'es',
    'son', 'del', 'por', 'para', 'como',
    'con', 'una', 'este', 'esta', 'pero',
  ],
  fr: [
    'le', 'la', 'les', 'est', 'sont',
    'des', 'une', 'avec', 'pour', 'dans',
    'pas', 'que', 'sur', 'mais', 'nous',
  ],
  de: [
    'der', 'die', 'das', 'ist', 'sind',
    'ein', 'eine', 'und', 'mit', 'nicht',
    'auf', 'ich', 'sie', 'den', 'auch',
  ],
};

interface ScriptRange {
  lang: string;
  ranges: [number, number][];
}

const SCRIPT_RANGES: ScriptRange[] = [
  {
    lang: 'ko',
    ranges: [
      [0xac00, 0xd7af],
      [0x1100, 0x11ff],
      [0x3130, 0x318f],
    ],
  },
  {
    lang: 'ja',
    ranges: [
      [0x3040, 0x309f],
      [0x30a0, 0x30ff],
    ],
  },
  {
    lang: 'zh',
    ranges: [
      [0x4e00, 0x9fff],
      [0x3400, 0x4dbf],
    ],
  },
  {
    lang: 'ar',
    ranges: [[0x0600, 0x06ff]],
  },
  {
    lang: 'ru',
    ranges: [[0x0400, 0x04ff]],
  },
  {
    lang: 'th',
    ranges: [[0x0e00, 0x0e7f]],
  },
  {
    lang: 'hi',
    ranges: [[0x0900, 0x097f]],
  },
];

function detectByScript(text: string): string | null {
  const counts: Record<string, number> = {};
  for (const ch of text) {
    const code = ch.codePointAt(0)!;
    for (const sr of SCRIPT_RANGES) {
      for (const [lo, hi] of sr.ranges) {
        if (code >= lo && code <= hi) {
          counts[sr.lang] =
            (counts[sr.lang] ?? 0) + 1;
        }
      }
    }
  }
  let best: string | null = null;
  let bestCount = 0;
  for (const [lang, count] of Object.entries(counts)) {
    if (count > bestCount) {
      best = lang;
      bestCount = count;
    }
  }
  return bestCount >= 2 ? best : null;
}

function detectByWords(text: string): string | null {
  const words = text
    .toLowerCase()
    .split(/\s+/)
    .filter((w) => w.length > 0);
  if (words.length === 0) return null;

  let best: string | null = null;
  let bestScore = 0;

  for (const [lang, common] of Object.entries(
    COMMON_WORDS,
  )) {
    let hits = 0;
    for (const w of words) {
      if (common.includes(w)) hits++;
    }
    const score = hits / words.length;
    if (score > bestScore) {
      best = lang;
      bestScore = score;
    }
  }

  return bestScore >= 0.1 ? best : null;
}

function detectLanguage(text: string): string | null {
  return detectByScript(text) ?? detectByWords(text);
}

export function language(
  options: LanguageOptions,
): Guard {
  return {
    name: 'language',
    version: '0.1.0',
    description: 'Language detection and filtering',
    category: 'locale',
    supportedStages: ['input', 'output'],
    async check(
      text: string,
      _ctx: GuardContext,
    ): Promise<GuardResult> {
      const start = performance.now();
      const detected = detectLanguage(text);
      const allowed =
        detected !== null &&
        options.allowed.includes(detected);
      const triggered = !allowed;

      return {
        guardName: 'language',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        latencyMs: Math.round(performance.now() - start),
        details: { detectedLanguage: detected },
      };
    },
  };
}

File: python/open_guardrail/guards/language.py
"""Language detection and filtering guard."""
from __future__ import annotations

import time
from typing import List, Optional

from open_guardrail.core import GuardResult

_SCRIPT_RANGES: dict[str, list[tuple[int, int]]] = {
    "ko": [(0xAC00, 0xD7AF), (0x3130, 0x318F), (0x1100, 0x11FF)],
    "ja": [(0x3040, 0x309F), (0x30A0, 0x30FF), (0x4E00, 0x9FFF)],
    "zh": [(0x4E00, 0x9FFF), (0x3400, 0x4DBF)],
    "ar": [(0x0600, 0x06FF), (0x0750, 0x077F)],
    "th": [(0x0E00, 0x0E7F)],
    "hi": [(0x0900, 0x097F)],
    "ru": [(0x0400, 0x04FF)],
}

_COMMON_WORDS: dict[str, list[str]] = {
    "en": ["the", "is", "are", "was", "were", "have", "has", "and", "for", "that", "this", "with"],
    "ko": ["이", "그", "저", "것", "수", "를", "에", "의", "은", "는", "가"],
    "ja": ["の", "は", "が", "を", "に", "で", "と", "も", "から", "まで"],
    "zh": ["的", "是", "在", "了", "不", "和", "有", "这", "人", "我"],
    "es": ["el", "la", "de", "en", "que", "los", "del", "las", "por", "con"],
    "fr": ["le", "la", "de", "et", "les", "des", "en", "un", "une", "est"],
    "de": ["der", "die", "und", "den", "von", "ist", "das", "ein", "mit", "auf"],
}


def _detect_language(text: str) -> Optional[str]:
    scores: dict[str, int] = {}
    for ch in text:
        cp = ord(ch)
        for lang, ranges in _SCRIPT_RANGES.items():
            for lo, hi in ranges:
                if lo <= cp <= hi:
                    scores[lang] = scores.get(lang, 0) + 1
    words = text.lower().split()
    for lang, word_list in _COMMON_WORDS.items():
        count = sum(1 for w in words if w in word_list)
        if count >= 2:
            scores[lang] = scores.get(lang, 0) + count * 3
    if not scores:
        return None
    return max(scores, key=lambda k: scores[k])


class _Language:
    def __init__(self, *, allowed: List[str], action: str = "block") -> None:
        self.name = "language"
        self.action = action
        self.allowed = [a.lower() for a in allowed]

    def check(self, text: str, stage: str = "input") -> GuardResult:
        start = time.perf_counter()
        detected = _detect_language(text)
        triggered = detected is not None and detected not in self.allowed
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name="language",
            passed=not triggered,
            action=self.action if triggered else "allow",
            message=f"Language '{detected}' not allowed" if triggered else None,
            latency_ms=round(elapsed, 2),
            details={"detected": detected, "allowed": self.allowed} if triggered else None,
        )


def language(*, allowed: List[str], action: str = "block") -> _Language:
    return _Language(allowed=allowed, action=action)
